packages/cadl-ranch-specs/http/authentication/apiKey/mockapi.ts

```diff
@@ -0,0 +1,25 @@
+import { passOnSuccess, mockapi, json } from "@azure-tools/cadl-ranch-api";
+import { ScenarioMockApi } from "@azure-tools/cadl-ranch-api";
+
+export const Scenarios: Record<string, ScenarioMockApi> = {};
+
+Scenarios.ApiKeyAuthentication_authenticated = passOnSuccess(
+  mockapi.get("/authentication/api-key/authenticated", (req) => {
+    req.expect.containsHeader("x-ms-api-key", "valid-key");
+    return { status: 204 };
+  }),
+);
+
+Scenarios.ApiKeyAuthentication_invalidAuthentication = passOnSuccess(
+  mockapi.get("/authentication/api-key/invalid-authentication", (req) => {
+    return {
+      status: 403,
```

Comment:
Should we have a scenario for 401 too if the header name is incorrect or missing?

Reply:
Would generator handle those differently?

Reply:
copy one comment [above](https://github.com/Azure/cadl-ranch/pull/56#discussion_r951378553)

> Since api-key is usually not via "Authentication" header, return 403 (instead of 401 which would require a WWW-Authenticate header in response for challenge).
>
> Checked storage and search, both uses 403 (though storage actually send the SharedKey via Authentication header).

`WWW-Authenticate` is required for 401, per [rfc7235](https://www.rfc-editor.org/rfc/rfc7235#section-4.1).
But so far I didn't find a case in our backend in ApiKey case (admit I am not an expert on auth, so maybe we need change this later).

E.g., if I send "https://management.azure.com/subscriptions/ec0aa5f7-9e78-40c9-85cd-535c6305b380/resourceGroups/rg-weidxu?api-version=2021-01-01" which require Bearer auth (the OAuth2 case), I will get

```
401

WWW-Authenticate: Bearer authorization_uri="https://login.windows.net/72f988bf-86f1-41af-91ab-2d7cd011db47", error="invalid_token", error_description="The authentication failed because of missing 'Authorization' header."

{
    "error": {
        "code": "AuthenticationFailed",
        "message": "Authentication failed. The 'Authorization' header is missing."
    }
}
```

---

Code generator does not handle it differently (as this is runtime behavior, not in swagger or cadl). Usually azure-core or azure-identity (on each language) would handle the ApiKey and OAuth2 flow.

The difference of 401 and 403 in runtime would likely get SDK client throw a different error/exception.

Reply:
don't really have a strong opinion on this, if you feel like having 401 and 403 is valuable I'm down to add it otherwise I think this PR Is good! We can always add the 401 scenario later if we find more value in it.

Reply:
So far, no opinion from me. 401 will likely happen to OAuth2 case.

@qiaozha @msyyc @pshao25 for opinion from SDK.

Reply:
In my understanding, we only need to test that expected key are set to header `x-ms-api-key`. We even don't the need the `ApiKeyAuthentication_invalidAuthentication` scenario, because it is not what a service need to handle.

When a service receives a request, it just redirects to an auth provider with the auth headers. If auth provider says it fails, service would send back the suth result to the client. 401/403 is what auth provider provides, we clients should not do anything about this.

Reply:
well however a service implement the authentication logic is irrelevant. The result from a client point of view is that it can succeed or fail. And it is definitely valuable to have a scenario where you check the SDK doesn't blow up with unhelpful information if the service return authentication failure.

Reply:
> In my understanding, we only need to test that expected key are set to header `x-ms-api-key`. We even don't need the `ApiKeyAuthentication_invalidAuthentication` scenario, because it is not what a service need to handle.
> 
> When a service receives a request, it just redirects to an auth provider with the auth headers. If auth provider says it fails, service would send back the auth result to the client. 401/403 is what auth provider provides, we clients should not do anything about this.
> 
> So my opinion is to delete this scenario.

No, 401/403 is what service respond. You can try with any of your existing SDK.

So far none of them is about auth server/provider.